### PR TITLE
Add -Z normalize-docs and enable it for compiler docs

### PR DESCRIPTION
**Pull request by jyn514**
Works around https://github.com/rust-lang/rust/issues/79459 by only enabling normalization for the compiler itself (and anyone who opts-in on nightly). Eventually I want to remove this and enable normalization by default, but that's turned out to be [really hard](https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/How.20do.20I.20normalize.20projection.20types.20to.20a.20single.20type.3F/near/218125195).

This uses a command line option instead of a feature gate so it's easier to pass it to all crates at once. Theoretically it's better to use a feature gate instead so that it's easier for people to use on docs.rs, but I'm also not terribly concerned with how easy it to use a temporary hack.

Addresses https://github.com/rust-lang/rust/issues/77459.

### CODE REVIEW

**Comment by jyn514** [comment]
(I'm testing locally that this actually works and generates the correct docs, this probably shouldn't be approved until then.)

**Comment by jyn514** [comment]
Update: it does not work.

**Comment by jyn514** [comment]
Ok, it's working now. Not sure if I should add a test that the option is unstable - it uses `-Z` so I'd assume it's just as unstable as every other option?

**Comment by GuillaumeGomez** [comment]
If you have to use `-Z unstable`, then it's unstable. Also, I'd like to have someone from the @rust-lang/compiler tool to check this PR as well to be sure we didn't forget anything.

**Comment by nagisa** [comment]
None of the `-Z` options actually require `-Zunstable-options`. `-Z` by itself is intended to signal the instability and stable compiler will blanket-reject any and all `-Z` options. The `-Zunstable-options` flag exists to enable making non-`-Z` flags (such as `--pretty`) unstable.

As long as the newly added flag is in the `DebuggingOptions` section you're good on stability.

File: src/librustdoc/clean/mod.rs [hunk 1]
```diff
@@ -1505,7 +1505,11 @@ impl Clean<Type> for hir::Ty<'_> {
 /// Returns `None` if the type could not be normalized
 #[allow(unreachable_code, unused_variables)]
```
Comment (SNCPlay42) [inline src/librustdoc/clean/mod.rs:1506]:
```suggestion
```
Making the return conditional should make everything reachable here.

Comment (jyn514) [reply src/librustdoc/clean/mod.rs]: Thanks, fixed!

**Comment by jyn514** [comment]
@bors r=GuillaumeGomez

We'll *finally* have compiler docs that look nice 🎉

**Pull request merged**